### PR TITLE
feat: INTMDB 732 atlas bootstrap

### PR DESCRIPTION
**Pull request by Adelmar92**
## Proposed changes

added CDK Bootstrap per https://github.com/mongodb/awscdk-resources-mongodbatlas/issues/11: 

The atlas-cdk-bootstrap AWS Cloud Development Kit (CDK) construct was designed to facilitate the smooth configuration and setup of the MongoDB Atlas CDK framework. This construct simplifies the process of preparing your environment to run the Atlas CDK by automating essential configurations and resource provisioning.

Key Features:

- User Provisioning: The atlas-cdk-bootstrap construct creates a dedicated user within AWS Identity and Access Management (IAM) for executing CloudFormation resources. This helps maintain security and isolation for Atlas CDK operations.

- API Key Management: It sets up an Amazon Secret Manager to securely store and manage API keys required for interacting with the Atlas services. This ensures sensitive credentials are protected and can be easily rotated.

- CloudFormation Resource Activation: This construct streamlines the activation of CloudFormation resources essential for Atlas CDK. It provides a seamless interface for users to specify the specific CloudFormation resources that need to be deployed and configured.

With atlas-cdk-bootstrap, you can accelerate the onboarding process for Atlas CDK and significantly reduce the complexity of environment setup. By automating user provisioning, credential management, and resource activation, this CDK construct empowers developers to focus on building and deploying applications using Atlas CDK without getting bogged down by manual configuration tasks.

Whether you're a seasoned developer or just starting with Atlas CDK, atlas-cdk-bootstrap offers a convenient and efficient way to kickstart your projects, enhance security, and ensure a smooth development experience. Streamline your Atlas CDK setup today and unlock the true potential of cloud application development.

Testing:
ive runned the deploy command:
image

the secret was created with the parameters:
image

the role is created with all its permissions:
image

the four resources get activated: 
image

Link to any related issue(s): 
[INTMDB-732](https://jira.mongodb.org/browse/INTMDB-732)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code
- [ ] I have tested the CDK constructor in a CFN stack. See [TESTING.md](../TESTING.md)

### CODE REVIEW

**Comment by Zuhairahmed** [comment]
Thanks @Adelmar92! @pahud feel to review as well, thanks again for feedback here in advance

**Comment by pahud** [comment]
Thank you. This is awesome!

As the private key is the most sensitive data, we should not expose that anywhere even in the cloudformation console.

If we deploy the bootstrap stack by passing the private key through `--parameters`, the value of the keys would be exposed in the CFN console which means anyone who can access the CFN console can read the key values.

I would suggest:

1. the `bootstrap` stack should first create a `dummy` secret to hold the key pairs.
2. instead of passing the `--parameters`, which holds and exposes the secret values in the CFN console, users will be required to run "AWS CLI" to update the dummy secret. This will ensure the secrets will be only visible to those who have the IAM access to that secret.


```
$ export MONGO_ATLAS_PUBLIC_KEY='your_public_key'
$ export MONGO_ATLAS_PRIVATE_KEY='your_private_key'
# update the secret with AWS CLI
$ aws secretsmanager update-secret --secret-id cfn/atlas/profile/my-mongo-profile --secret-string "{\"PublicKey\":\"${MONGO_ATLAS_PUBLIC_KEY}\",\"PrivateKey\":\"${MONGO_ATLAS_PRIVATE_KEY}\"}"
```

What do you think?

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        "MongoSecretProfile",
+        props?.secretProfile,
+        props.atlasPublicKey,
+        props.atlasPrivateKey
+      );
+    }
+
+    if (props?.typesToActivate) {
+      for (let type of props?.typesToActivate) {
+        new cloudformation.CfnTypeActivation(this, `${type}TypeActivation`, {
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/index.ts:121]:
```suggestion
        new cloudformation.CfnTypeActivation(this, `${type}TypeActivation`, {
```

Good to know this 👍

**Comment by Adelmar92** [comment]
> Thank you. This is awesome!
> 
> As the private key is the most sensitive data, we should not expose that anywhere even in the cloudformation console.
> 
> If we deploy the bootstrap stack by passing the private key through `--parameters`, the value of the keys would be exposed in the CFN console which means anyone who can access the CFN console can read the key values.
> 
> I would suggest:
> 
> 1. the `bootstrap` stack should first create a `dummy` secret to hold the key pairs.
> 2. instead of passing the `--parameters`, which holds and exposes the secret values in the CFN console, users will be required to run "AWS CLI" to update the dummy secret. This ensures the secrets will only be visible to those who have the IAM access to that secret.
> 
> ```
> $ export MONGO_ATLAS_PUBLIC_KEY='your_public_key'
> $ export MONGO_ATLAS_PRIVATE_KEY='your_private_key'
> # update the secret with AWS CLI
> $ aws secretsmanager update-secret --secret-id cfn/atlas/profile/my-mongo-profile --secret-string "{\"PublicKey\":\"${MONGO_ATLAS_PUBLIC_KEY}\",\"PrivateKey\":\"${MONGO_ATLAS_PRIVATE_KEY}\"}"
> ```
> 
> What do you think?

Agree with this one Pahud! im applying this change

**Comment by pahud** [comment]
> Hi Pahud in this case you suggest adding the bootstap in conjuction with other l3? like atlas basic?

Because the `bootstrap` is a custom construct that prepares the initial CFN extension execution role as well as activate types via `CfnTypeActivation`, it makes sense to make it as a separate construct in `awscdk-resources-mongodbatlas`.

If it's the first time for the user to create MongoDB Atlas resources with CDK, they will have to

```ts
import * as atlas from 'awscdk-resources-mongodbatlas';
import { App, Stack } from 'aws-cdk-lib'

const app = new App();
const bootstrapStack = new Stack(app, 'mongodb-bootstrap');
new atlas.MongoAtlasBootstrap(bootstrapStack, 'mongodb-bootstrap');
```
And run `npx cdk deploy mongodb-bootstrap` to deploy the bootstrap stack.

We should encourage user to create AtlasBasic in a separate stack like:

```ts
import * as atlas from 'awscdk-resources-mongodbatlas';
import { App, Stack } from 'aws-cdk-lib'

const app = new App();
const bootstrapStack = new Stack(app, 'mongodb-bootstrap');
const clusterStack = new Stack(app, 'mongodb-cluster');
// Create the bootstrap resources in the bootstrap stack. 
new atlas.MongoAtlasBootstrap(bootstrapStack, 'mongodb-bootstrap');
// Create the AtlasCluster in another stack
new atlas.AtlasBasic(clusterStack, 'cluster-stack', props);
```

The idea is that we should encourage users to leave the bootstrap there even when they destroy the AtlasCluster because the bootstrap only need to be created once.

However, if users would really like to clean up literally EVERYTHING they still can `npx cdk destroy mongodb-bootstrap` and the cfn extension execution role, secret manager secret as well as all the CfnTypeActivations will be removed.

What do you think?

**Comment by Zuhairahmed** [comment]
just quick update @pahud , this item is on our roadmap and hopefully looking to close out in a few weeks. stay tuned!

**Comment by Adelmar92** [comment]
> > Hi Pahud in this case you suggest adding the bootstap in conjuction with other l3? like atlas basic?
> 
> Because the `bootstrap` is a custom construct that prepares the initial CFN extension execution role as well as activate types via `CfnTypeActivation`, it makes sense to make it as a separate construct in `awscdk-resources-mongodbatlas`.
> 
> If it's the first time for the user to create MongoDB Atlas resources with CDK, they will have to
> 
> ```ts
> import * as atlas from 'awscdk-resources-mongodbatlas';
> import { App, Stack } from 'aws-cdk-lib'
> 
> const app = new App();
> const bootstrapStack = new Stack(app, 'mongodb-bootstrap');
> new atlas.MongoAtlasBootstrap(bootstrapStack, 'mongodb-bootstrap');
> ```
> 
> And run `npx cdk deploy mongodb-bootstrap` to deploy the bootstrap stack, which contains everything they need before creating their first cluster.
> 
> We should encourage user to create AtlasBasic in a separate stack like:
> 
> ```ts
> import * as atlas from 'awscdk-resources-mongodbatlas';
> import { App, Stack } from 'aws-cdk-lib'
> 
> const app = new App();
> const bootstrapStack = new Stack(app, 'mongodb-bootstrap');
> const clusterStack = new Stack(app, 'mongodb-cluster');
> // Create the bootstrap resources in the bootstrap stack. 
> new atlas.MongoAtlasBootstrap(bootstrapStack, 'mongodb-bootstrap');
> // Create the AtlasCluster in another stack
> new atlas.AtlasBasic(clusterStack, 'cluster-stack', props);
> ```
> 
> We encourage users to retain the bootstrap stack even when they destroy the AtlasCluster because the bootstrap only requires to be created just once.
> 
> However, if users would really like to clean up literally EVERYTHING they still can `npx cdk destroy mongodb-bootstrap` and the cfn extension execution role, secret manager secret as well as all the CfnTypeActivations will be removed.
> 
> As a constant MongoDB Atlas user in AWS, I would retain the bootstrap stack and even enable the [stack termination protection](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-protect-stacks.html). However, for billing-sensitive users, the only billing incurred in the bootstrap stack should be the secret manager secret([$0.40 per secret per month](https://aws.amazon.com/secrets-manager/pricing/)), if they really don't need that anymore, they can always cdk destroy the bootstrap stack anytime then no billing will be incurred after that.
> 
> What do you think?

hello @pahud this make sense, but what should change from the current implementation, right now is defined as an L3 construct, in a separated construct. you can see that in this example https://github.com/mongodb/awscdk-resources-mongodbatlas/blob/feature/INTMDB-732-atlas-bootstrap/examples/l3-resources/atlas-bootstrap.ts, this should be implemented in a diferent whay?

**Review by andreaangiolillo — DISMISSED** [review]
LGTM

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+export class MongoSecretProfile extends Construct {
+  constructor(scope: Construct, id: string, profileName: string) {
+    super(scope, id);
+
+    // create a secret
+    const secret = new secretsmanager.Secret(this, profileName, {
+      secretName: `cfn/atlas/profile/${profileName}`,
+      description: "Secret used for MongoDB Atlas Cloud Formation api keys",
+      secretStringValue: SecretValue.unsafePlainText(
+        '{"PublicKey":"yourAtlasPublicKey", "PrivateKey": "yourAtlasPrivateKey"}'
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/index.ts:118]: I think we should randomize at least the private key on secret initialization.

```ts
new secretsmanager.Secret(this, id, {
      generateSecretString: {
        secretStringTemplate: JSON.stringify({ PublicKey: 'yourAtlasPublicKey', PrivateKey: '' }),
        generateStringKey: 'PrivateKey',
        excludeCharacters: '%+~`#$&*()|[]{}:;<>?!\'/@"\\=-.,',
      },
    });
```

And you got the initial secret like this:

image

Thoughts?

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ * Generate the CFN extension execution role.
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry-public.html
+ */
+export class MongoAtlasBootstrap extends Construct {
+  readonly role: iam.IRole;
+  constructor(scope: Construct, id: string, props?: MongoAtlasBootstrapProps) {
+    super(scope, id);
+
+    this.role = new iam.Role(this, "mongoDBAtlasCFNExecutionRole", {
+      maxSessionDuration: Duration.seconds(8400),
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/index.ts:50]: What will happen if we leave `maxSessionDuration` undefined?

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            "iam:GetRolePolicy",
+            "iam:ListAttachedRolePolicies",
+            "iam:ListRolePolicies",
+            "iam:PutRolePolicy",
+          ],
+          resources: ["*"],
+        }),
+      ],
+    });
+
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/index.ts:89]: We probably still need to scope down with this as described in the doc using [addSourceAccountCondition()](https://github.com/aws/aws-cdk/blob/c445b8cc6e20d17e4a536f17262646b291a0fe36/packages/aws-cdk-lib/aws-iam/lib/policy-statement.ts#L422).

```
"StringEquals":{
                    "aws:SourceAccount":"123456789012"
                },
```

File: examples/l3-resources/atlas-bootstrap.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    const roleName = "MongoDB-Atlas-CDK-Excecution";
+
+    const publicKey = new cdk.CfnParameter(this, "atlasPublicKey", {
+    noEcho: true,
+    type: "String",
+    });
+    const privateKey = new cdk.CfnParameter(this, "atlasPrivateKey", {
+    noEcho: true,
+    type: "String",
+    });
```
Comment (pahud) [inline examples/l3-resources/atlas-bootstrap.ts:37]: I don't think we should allow user to pass private key through the parameter. This means people with cloudformation stack describe privilege will be able to see the plaintext of the private key. If you deploy this and go to the cloudformation console, you should be able to see all the key values. Maybe we should avoid that.

File: src/l3-resources/atlas-bootstrap/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+[aws cloudformation registry]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry.html
+
+## Description
+
+The atlas-cdk-bootstrap AWS Cloud Development Kit (CDK) construct was designed to facilitate the smooth configuration and setup of the MongoDB Atlas CDK framework. This construct simplifies the process of preparing your environment to run the Atlas CDK by automating essential configurations and resource provisioning.
+
+Key Features:
+
+- User Provisioning: The atlas-cdk-bootstrap construct creates a dedicated user within AWS Identity and Access Management (IAM) for executing CloudFormation resources. This helps maintain security and isolation for Atlas CDK operations.
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/README.md:13]:
```suggestion
- User Provisioning: The atlas-cdk-bootstrap construct creates a dedicated execution role within AWS Identity and Access Management (IAM) for executing CloudFormation Extension resources. This helps maintain security and isolation for Atlas CDK operations.
```

File: src/l3-resources/atlas-bootstrap/README.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+- API Key Management: It sets up an Amazon Secret Manager to securely store and manage API keys required for interacting with the Atlas services. This ensures sensitive credentials are protected and can be easily rotated.
+
+- CloudFormation Resource Activation: This construct streamlines the activation of CloudFormation resources essential for Atlas CDK. It provides a seamless interface for users to specify the specific CloudFormation resources that need to be deployed and configured.
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/README.md:17]:
```suggestion
- CloudFormation Extensions Activation: This construct streamlines the activation of CloudFormation public extensions essential for Atlas CDK. It provides a seamless interface for users to specify the specific CloudFormation resources that need to be deployed and configured.
```

According to the [doc](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry-public.html), I'd prefer to call it "activate the CloudFormation Extentions".

File: src/l3-resources/atlas-bootstrap/README.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+With atlas-cdk-bootstrap, you can accelerate the onboarding process for Atlas CDK and significantly reduce the complexity of environment setup. By automating user provisioning, credential management, and resource activation, this CDK construct empowers developers to focus on building and deploying applications using Atlas CDK without getting bogged down by manual configuration tasks.
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/README.md:19]: I really love this statement!

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  /**
+   * The secret profile name for MongoDB Atlas.
+   * @default generate a dummy secret.
+   * @see https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master#2-configure-your-profile
+   */
+  readonly secretProfile!: string;
+  /**
+   * List of strings representing mongoDB atlas types to activate.
+   * the exported AtlasBasicResources can be used for AtlasBasic resources
+   * if not provided non resources will be activated
```
Comment (pahud) [inline src/l3-resources/atlas-bootstrap/index.ts:34]: Is there any use case that user would not activate any atlas basic resources?

Maybe when it's undefined, the default behavior is to activate all the basic resources instead? wdyt?

File: test/l3-resources/atlas-bootstrap/README.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# @awscdk-resources-mongodbatlas/atlas-basic
+
+## CFN resource type used
+
+- AWS::IAM::Role
+- AWS::IAM::Policy
+- AWS::CloudFormation::TypeActivation
+- AWS::SecretsManager::Secret
+
+These CFN resources must be active in your AWS account while using this constructor.
```
Comment (pahud) [inline test/l3-resources/atlas-bootstrap/README.md:10]: I didn't get it.

All the `AWS::*` resources are built-in and don't need any activation. Why should we activate them?

**Review by pahud — COMMENTED** [review]
Great PR. Just left some comments.

File: src/l3-resources/atlas-bootstrap/index.ts [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  readonly role: iam.IRole;
+  constructor(scope: Construct, id: string, props?: MongoAtlasBootstrapProps) {
+    super(scope, id);
+
+    this.role = new iam.Role(this, "mongoDBAtlasCFNExecutionRole", {
+      maxSessionDuration: Duration.seconds(8400),
+      assumedBy: new iam.CompositePrincipal(
+        new iam.ServicePrincipal("cloudformation.amazonaws.com"),
+        new iam.ServicePrincipal("resources.cloudformation.amazonaws.com"),
+        new iam.ServicePrincipal("lambda.amazonaws.com")
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply src/l3-resources/atlas-bootstrap/index.ts:54]: in our testings, we we got system failures in the cfn stack when all of them were not added in the Iam role

File: src/l3-resources/atlas-bootstrap/index.ts (hunk 3 above)
Comment (Adelmar92) [reply src/l3-resources/atlas-bootstrap/index.ts:50]: This comes as a default property when you create an execution role with CFN, i just replicated a role exactly as they are created by the CFN tool

File: examples/l3-resources/atlas-bootstrap.ts (hunk 5 above)
Comment (Adelmar92) [reply examples/l3-resources/atlas-bootstrap.ts]: this are gone, we dont need them anymore, ill remove them

File: src/l3-resources/atlas-bootstrap/index.ts (hunk 9 above)
Comment (Adelmar92) [reply src/l3-resources/atlas-bootstrap/index.ts:34]: mybe they have already activated them, or they want to manage them manually, so i left the option not to activate anything, or mybe choose witch resource they want to activate, mybe in the future they might want to modify the stack to add new ones

File: test/l3-resources/atlas-bootstrap/README.md (hunk 10 above)
Comment (Adelmar92) [reply test/l3-resources/atlas-bootstrap/README.md:10]: i found that AWS::CloudFormation::TypeActivation is not acivated for some reason in a few regions, this comment is just for the developers, to know which resources should be activated if thew want to test, but dont worry about this one, is just for developer testing, the user should not use this

File: src/l3-resources/atlas-bootstrap/index.ts (hunk 9 above)
Comment (Adelmar92) [reply src/l3-resources/atlas-bootstrap/index.ts:34]: also i see there is a few regions that does not have the AWS::CloudFormation::TypeActivation  me-central-1, ap-south-2, ap-southeast-4, eu-south-2, eu-central-2, so i just let the users decide not to activate them y the script, but do it manually

**Comment by pahud** [comment]
I didn't see any major issues now. This LGTM. Let's ship it.

**Review by Zuhairahmed — APPROVED** [review]
Let's merge, great work here @Adelmar92 !

**Pull request merged**